content/en/blog/attribute-autocomplete-enhancement/index.md

```diff
@@ -0,0 +1,306 @@
+---
+title: "Enhancing Attribute Autocomplete Functionality -  Auto Correction and Type Suggestion"
+description: ""
+lead: "We discuss the two new enhancement added to the attribute name autocomplete feature in Texera, auto attribute correction and attribute type suggestion"
+date: 2023-06-18T20:58:20-07:00
+lastmod: 2023-06-18T20:58:20-07:00
+draft: true
+weight: 50
+#images: ["attribute-autocomplete-enhancement.jpg"]
+contributors: ["Chengxi Li"]
+---
+
+# Enhancing Attribute Autocomplete Functionality -  Auto Correction and Type Suggestion
+
+In this blog, we discuss the two new enhancement added to the attribute name autocomplete feature in Texera, auto attribute correction and attribute type suggestion, making the feature more user-friendly.
+
+## Background
+
+### What are operator properties
+
+Texera workflows consist of sequences of operators. Every operator has properties as the parameters of the data process function. For example, in Fig. 1, the "Filter" operator has three properties, `Attribute`, `Condition`, and `Value`. The property values in Fig. 1 indicate that the operator filters the records with amount greater than 1.
+
+**![](https://lh6.googleusercontent.com/lxGfMeDfnw14f4NskeiSVBlz3_CFUQYrQJ8l97GnDo5L5LXoZ4w8mxJUa9AnnOWn7FLlxO-kwmValePube-Hw21BlwuhzFLco_pPhdrJu7SHObcQXJEtug4h-oBDIqnXdGXfwYlerpGvkZyaZ84bK4s)**
+
+Fig. 1, Filter operator properties
+
+### What is operator input schema
+
+Various formats of data can be processed by Texera, and the format of data is represented by its schema, which is a set of attributes. In order to let every operator know what attributes of the data are available, each operator deduces its output schema using its input schema and propagates it to the next operator. This mechanism is called "schema propagation." Schema propagation is an important part of the system, as it is the base of many useful features such as operator validation and attribute autocomplete.
+
+### What is attribute autocomplete
+
+Many operators have properties that take an attribute name of the input data as value. For example, the `Attribute` property in the "Sentiment Analysis" operator takes the name of the attribute on which it will perform analysis. Autocomplete feature allows users to select an input attribute from a dropdown menu instead of manually typing in the name.
+
+**![](https://lh3.googleusercontent.com/vqgcIM-ae7mRCIbah-ZyZRJkM4T4MMNdo3HGQtVT0RFfeh9J69z7SIUspMfY1IUjJ0V30U84PLdBMd3QouXKLEbfjUBc3Ac5q3_CtKk5XAP55I3QY0oG2U6Q11TIWcXUrVfjhbGpUUgV7LPp1R4ACeM)**
+
+Fig. 2, Dropdown menu for attribute autocomplete
+
+## Auto Attribute Correction
```

Comment:
I feel `Attribute Correction` is not a good name. It removes an invalid value from an attribute property, but `correction` is implying you can fix the invalid value to a valid one. Maybe call it `Attribute Sanitization`?

Reply:
I think it does fix an invalid name to a valid one; for example, when an attribute selection becomes invalid because its name is changed, we correct it to the new valid name. Thus the service is called `AutoAttributeCorrectionService`.

Reply:
ok.